### PR TITLE
FEAT-#1925: add import from arrow table

### LINKED ISSUE #1925: [FEAT] Add import from arrow table

### PR DESCRIPTION
**Pull request by ienkovich**
## What do these changes do?
Add API to build modin dataframes from arrow tables. This is an important feature for arrow based OmniSci backend. For pandas based backend we simple convert arrow table to pandas dataframe.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1925   
- [x] tests added and passing

### CODE REVIEW

File: modin/backends/base/query_compiler.py [hunk 1]
```diff
@@ -124,6 +124,28 @@ def from_pandas(cls, df, data_cls):
...
 
+    # From Arrow
+    @classmethod
+    @abc.abstractmethod
+    def from_arrow(cls, at, data_cls):
+        """Improve simple Arrow Table to an advanced and superior Modin DataFrame.
+
+        Parameters
+        ----------
+        df: Arrow Table
```
Comment (YarShev) [inline modin/backends/base/query_compiler.py:135]:
```suggestion
        at : Arrow Table
```

File: modin/engines/base/frame/data.py [hunk 2]
```diff
@@ -1732,6 +1732,39 @@ def from_pandas(cls, df):
             dtypes=new_dtypes,
         )
 
+    @classmethod
+    def from_arrow(cls, at):
+        """Improve simple Arrow Table to an advanced and superior Modin DataFrame.
+
+        Args:
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1739]:
```suggestion
        Parameters
        ----------
```

File: modin/engines/base/frame/data.py [hunk 3]
```diff
@@ -1732,6 +1732,39 @@ def from_pandas(cls, df):
...
+            at: Arrow Table object.
+
+        Returns:
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1742]:
```suggestion
        Returns
        -------
        BasePandasFrame
```

File: modin/engines/base/frame/data.py [hunk 4]
```diff
@@ -1732,6 +1732,39 @@ def from_pandas(cls, df):
...
+            A new dataframe.
+        """
+        new_frame, new_lengths, new_widths = cls._frame_mgr_cls.from_arrow(at, True)
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1745]:
```suggestion
        new_frame, new_lengths, new_widths = cls._frame_mgr_cls.from_arrow(at, return_dims=True)
```

File: modin/engines/base/frame/data.py [hunk 5]
```diff
@@ -1732,6 +1732,39 @@ def from_pandas(cls, df):
...
+            columns=new_columns,
+            row_lengths=new_lengths,
+            column_widths=new_widths,
+            dtypes=new_dtypes,
+        )
+
+    @classmethod
+    def _arrow_type_to_dtype(cls, arrow_type):
+        res = arrow_type.to_pandas_dtype()
+        if not isinstance(res, (np.dtype, str)):
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1764]: Why is this condition needed?

**Comment by YarShev** [comment]
@ienkovich , can you please create an issue for the PR?

File: modin/engines/base/frame/data.py (hunk 5 above)
Comment (ienkovich) [reply modin/engines/base/frame/data.py:1766]: Arrow function `to_pandas_dtype` may return both dtype objects and numpy types, this is to unify returned value type for simpler processing.

**Comment by ienkovich** [comment]
Thanks for your comments. I created a new issue for this patch and pushed the requested changes.

File: modin/engines/base/frame/data.py [hunk 6]
```diff
@@ -1732,6 +1732,41 @@ def from_pandas(cls, df):
             dtypes=new_dtypes,
         )
 
+    @classmethod
+    def from_arrow(cls, at):
+        """Improve simple Arrow Table to an advanced and superior Modin DataFrame.
+
+        Parameters
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1739]:
```suggestion
        Parameters
        ----------
```

File: modin/engines/base/frame/data.py [hunk 7]
```diff
@@ -1732,6 +1732,41 @@ def from_pandas(cls, df):
...
+            at: Arrow Table object.
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1740]:
```suggestion
            at : Arrow Table
                The Arrow Table to convert from
```

File: modin/backends/base/query_compiler.py [hunk 8]
```diff
@@ -124,6 +124,28 @@ def from_pandas(cls, df, data_cls):
...
 
+    # From Arrow
+    @classmethod
+    @abc.abstractmethod
+    def from_arrow(cls, at, data_cls):
+        """Improve simple Arrow Table to an advanced and superior Modin DataFrame.
+
+        Parameters
+        ----------
+        at: Arrow Table
```
Comment (YarShev) [inline modin/backends/base/query_compiler.py:135]:
```suggestion
        at : Arrow Table
```

File: modin/engines/base/frame/data.py [hunk 9]
```diff
@@ -1732,6 +1732,41 @@ def from_pandas(cls, df):
...
+
+        Returns
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1742]:
```suggestion
        Returns
        -------
        BasePandasFrame
```

File: modin/engines/base/frame/data.py (hunk 5 above)
Comment (YarShev) [reply modin/engines/base/frame/data.py:1766]: I see, thanks!

File: modin/engines/base/frame/data.py [hunk 10]
```diff
@@ -1732,6 +1732,41 @@ def from_pandas(cls, df):
...
+            A new dataframe.
+        """
+        new_frame, new_lengths, new_widths = cls._frame_mgr_cls.from_arrow(
+            at, return_dims=True
+        )
+        new_columns = Index.__new__(Index, data=at.column_names, dtype="O")
+        new_index = Index.__new__(RangeIndex, data=range(at.num_rows))
+        new_dtypes = pandas.Series(
+            [cls._arrow_type_to_dtype(i.type) for i in at.columns],
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1751]:
```suggestion
            [cls._arrow_type_to_dtype(col.type) for col in at.columns],
```

File: modin/pandas/utils.py [hunk 11]
```diff
@@ -39,6 +39,20 @@ def from_pandas(df):
     return DataFrame(query_compiler=EngineDispatcher.from_pandas(df))
 
 
+def from_arrow(at):
+    """Converts an arrow Table to a Modin DataFrame.
+    Args:
```
Comment (YarShev) [inline modin/pandas/utils.py:44]:
```suggestion
    Parameters
    ----------
```

File: modin/pandas/utils.py [hunk 12]
```diff
@@ -39,6 +39,20 @@ def from_pandas(df):
...
+        at (pyarrow table): The Arrow table to convert.
+
+    Returns:
```
Comment (YarShev) [inline modin/pandas/utils.py:47]:
```suggestion
    Returns
    -------
    DataFrame
```

File: modin/pandas/utils.py [hunk 13]
```diff
@@ -39,6 +39,20 @@ def from_pandas(df):
     return DataFrame(query_compiler=EngineDispatcher.from_pandas(df))
 
 
+def from_arrow(at):
+    """Converts an arrow Table to a Modin DataFrame.
+    Args:
+        at (pyarrow table): The Arrow table to convert.
```
Comment (YarShev) [inline modin/pandas/utils.py:45]:
```suggestion
        at (pyarrow table) : Arrow Table
            The Arrow Table to convert from
```

File: modin/pandas/utils.py [hunk 14]
```diff
@@ -39,6 +39,20 @@ def from_pandas(df):
     return DataFrame(query_compiler=EngineDispatcher.from_pandas(df))
 
 
+def from_arrow(at):
+    """Converts an arrow Table to a Modin DataFrame.
```
Comment (YarShev) [inline modin/pandas/utils.py:43]:
```suggestion
    """Converts an Arrow Table to a Modin DataFrame.
```

**Comment by YarShev** [comment]
Also, I think the commit message has to be the following: `FEAT-#1925: add import from arrow table` because the PR resolves #1925, not #1871.

**Comment by ienkovich** [comment]
Thanks for your review! I applied the proposed changes in docstrings.

File: modin/engines/base/frame/data.py [hunk 15]
```diff
@@ -1732,6 +1732,44 @@ def from_pandas(cls, df):
...
+    def from_arrow(cls, at):
+        """Improve simple Arrow Table to an advanced and superior Modin DataFrame.
+
+        Parameters
+        ----------
+            at : Arrow Table
+                The Arrow Table to convert from.
+
+        Returns
+        -------
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1745]:
```suggestion
        -------
        BasePandasFrame
```

File: modin/pandas/utils.py [hunk 16]
```diff
@@ -39,6 +39,23 @@ def from_pandas(df):
     return DataFrame(query_compiler=EngineDispatcher.from_pandas(df))
 
 
+def from_arrow(at):
+    """Converts an Arrow Table to a Modin DataFrame.
+    Parameters
```
Comment (YarShev) [inline modin/pandas/utils.py:44]:
```suggestion

    Parameters
```

File: modin/engines/base/frame/data.py (hunk 15 above)
Comment (YarShev) [reply modin/engines/base/frame/data.py:1745]: `BasePandasFrame` should be added as returned type.

File: modin/pandas/utils.py (hunk 16 above)
Comment (YarShev) [reply modin/pandas/utils.py:44]: Empty line should be added between lines 43 and 44.

File: modin/pandas/utils.py [hunk 17]
```diff
@@ -39,6 +39,23 @@ def from_pandas(df):
...
+    ----------
+        at : Arrow Table
+            The Arrow Table to convert from.
+
+    Returns
+    -------
```
Comment (YarShev) [inline modin/pandas/utils.py:50]:
```suggestion
    -------
    DataFrame
```

Comment (YarShev) [reply modin/pandas/utils.py:50]: `DataFrame` should be added as returned type.

**Comment by ienkovich** [comment]
Thanks for your patient review. Hope this time all docstrings are well-formated. Unfortunately, there are many inconsistences in our docstring formatting. Following some random example from the current code didn't work well for me.

**Comment by YarShev** [comment]
@ienkovich , thanks for the changes, LGTM! Yes, we have many inconsistencies in our docstrings. We will worth fixing these in the future.

**Pull request merged**